File: scripts/paste_verdict.py
#!/usr/bin/env python3
"""Format a CardSnap scan JSON payload as a proof-paste reply (no link)."""

from __future__ import annotations

import json
import sys


def fmt_usd(n: float | int | None) -> str:
    if n is None:
        return "—"
    return f"${round(float(n)):,}"


def fmt_signed(n: float | int) -> str:
    v = round(float(n))
    s = fmt_usd(abs(v))
    if v > 0:
        return f"+{s}"
    if v < 0:
        return f"−{s}"
    return s


def raw_mid(payload: dict) -> float:
    mid = payload.get("rawValueMid") or 0
    if mid:
        return float(mid)
    lo = float(payload.get("rawValueLow") or 0)
    hi = float(payload.get("rawValueHigh") or 0)
    return (lo + hi) / 2


def format_paste(payload: dict) -> str:
    roi = payload.get("roi") or {}
    name = payload.get("confirmedName") or "this card"
    psa9 = float(payload.get("gradedPSA9Value") or 0)
    psa10 = float(payload.get("gradedPSA10Value") or 0)
    raw = float(roi.get("rawLiquidationUsd") or raw_mid(payload))
    net9 = float(roi.get("netIfPSA9") or 0)
    net10 = float(roi.get("netIfPSA10") or 0)
    verdict = roi.get("headlineVerdict") or "skip"
    psa9_pain = bool(roi.get("psa9PainCase"))

    ebay = payload.get("ebay") or {}
    comp_source = ebay.get("compSource")
    has_ebay = ebay.get("avgSoldPrice") is not None
    if comp_source is None and has_ebay:
        comp_source = "ebay_active_listings"

    psa = payload.get("psa") or {}
    has_pop = psa.get("psa9Pop") is not None or psa.get("psa10Pop") is not None
    if has_ebay and has_pop:
        qualifier = "recent comps"
    elif has_ebay or has_pop:
        qualifier = "directional comps"
    else:
        qualifier = "thin comps / model estimate"

    lines = [
        f"Ran {name} — {qualifier}:",
        f"Raw ~{fmt_usd(raw)} · PSA 9 ~{fmt_usd(psa9)} (net ~{fmt_signed(net9)} after fees) · PSA 10 ~{fmt_usd(psa10)} (net ~{fmt_signed(net10)} after fees)",
    ]

    if verdict == "grade" and psa9_pain:
        lines.append(
            f"Headline assumes PSA 10 — a 9 lands around {fmt_signed(net9)} net after PSA fee + est. shipping."
        )
    elif verdict == "skip":
        lines.append(
            f"Verdict: skip — modeled upside on a 10 is {fmt_signed(net10)} vs selling raw."
        )

    if comp_source == "ebay_active_listings":
        lines.append("Raw comp uses active asking prices, not confirmed solds.")

    lines.append("Happy to re-run if you name a different card or condition.")
    return "\n".join(lines)


def main() -> int:
    raw = sys.stdin.read()
    if not raw.strip():
        print("Pass ScanResultPayload JSON on stdin", file=sys.stderr)
        return 1
    payload = json.loads(raw)
    print(format_paste(payload))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: lib/paste-verdict.ts
import type { ScanResultPayload } from "@/lib/types";
import { formatUsd, formatUsdSigned } from "@/lib/format-currency";
import { computeGradingRoi } from "@/lib/roi";
import { describeCompSource } from "@/lib/source-confidence";

export type CommunityPasteReadiness = {
  ok: boolean;
  reasons: string[];
  fixHints: string[];
};

function hasEbayPrice(avg: number | null | undefined): boolean {
  return avg != null && !Number.isNaN(avg) && avg > 0;
}

function identityTooVague(data: ScanResultPayload): boolean {
  const set = (data.set ?? "").trim().toLowerCase();
  const name = (data.confirmedName ?? "").trim().toLowerCase();
  return set === "unknown" || set === "" || /\bunknown\b/.test(name);
}

/** Block Reddit proof-paste when comps or card ID are not trustworthy. */
export function assessCommunityPasteReadiness(
  data: ScanResultPayload
): CommunityPasteReadiness {
  const roi = data.roi ?? computeGradingRoi(data, undefined, data);
  const reasons: string[] = [];
  const fixHints: string[] = [];

  if (!hasEbayPrice(data.ebay.avgSoldPrice)) {
    reasons.push("No live marketplace comps for this search.");
    fixHints.push(
      "Re-run with year, set, card number, serial (/5, 1/1), and parallel name from the slab."
    );
  }

  if (identityTooVague(data)) {
    reasons.push("Card identity is too vague (unknown set or name).");
    fixHints.push("Copy the exact text from the BGS/PSA label or card back.");
  }

  if (roi.rawLiquidationUsd <= 0 && data.gradedPSA9Value <= 0 && data.gradedPSA10Value <= 0) {
    reasons.push("Modeled values are all $0 — verdict is not usable.");
  }

  const cardMatch = roi.loseMoneyReasons.find((r) => r.startsWith("Card match:"));
  if (cardMatch) {
    reasons.push(cardMatch.replace(/^Card match:\s*/, ""));
  }

  return {
    ok: reasons.length === 0,
    reasons,
    fixHints: Array.from(new Set(fixHints)),
  };
}

export function formatPasteVerdict(data: ScanResultPayload): string {
  const roi = data.roi ?? computeGradingRoi(data, undefined, data);
  const comp = describeCompSource(data);
  const qualifier =
    comp.overall === "High"
      ? "recent comps"
      : comp.overall === "Medium"
        ? "directional comps"
        : "thin comps / model estimate";

  const lines = [
    `Ran ${data.confirmedName} — ${qualifier}:`,
    `Raw ~${formatUsd(roi.rawLiquidationUsd)} · PSA 9 ~${formatUsd(data.gradedPSA9Value)} (net ~${formatUsdSigned(roi.netIfPSA9)} after fees) · PSA 10 ~${formatUsd(data.gradedPSA10Value)} (net ~${formatUsdSigned(roi.netIfPSA10)} after fees)`,
  ];

  if (roi.headlineVerdict === "grade" && roi.psa9PainCase) {
    lines.push(
      `Headline assumes PSA 10 — a 9 lands around ${formatUsdSigned(roi.netIfPSA9)} net after PSA fee + est. shipping.`
    );
  } else if (roi.headlineVerdict === "skip") {
    lines.push(
      `Verdict: skip — modeled upside on a 10 is ${formatUsdSigned(roi.netIfPSA10)} vs selling raw.`
    );
  }

  const cardMatchLine = roi.loseMoneyReasons.find((r) => r.startsWith("Card match:"));
  if (cardMatchLine) {
    lines.push(cardMatchLine.replace(/^Card match:\s*/, "⚠️ "));
  }

  if (comp.rawLabel.includes("Active listings")) {
    lines.push("Raw comp uses active asking prices, not confirmed solds.");
  }

  lines.push("Happy to re-run if you name a different card or condition.");

  return lines.join("\n");
}
